src/appleseed/renderer/modeling/aov/diagnosticaov.cpp

```diff
@@ -0,0 +1,339 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "diagnosticaov.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/aov/aovaccumulator.h"
+#include "renderer/kernel/aov/imagestack.h"
+#include "renderer/kernel/rendering/pixelcontext.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/kernel/shading/shadingresult.h"
+#include "renderer/modeling/aov/aov.h"
+#include "renderer/modeling/frame/frame.h"
+
+// appleseed.foundation headers.
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/image/tile.h"
+#include "foundation/utility/api/apistring.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+
+// Standard headers.
+#include <cstddef>
+#include <string>
+
+using namespace foundation;
+using namespace std;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Diagnostic AOV accumulator.
+    //
+
+    class DiagnosticAOVAccumulator
+      : public AOVAccumulator
+    {
+      public:
+        DiagnosticAOVAccumulator(
+            Image& image,
+            const string& diagnostic_name)
+          : m_image(image)
+          , m_diagnostic_name(diagnostic_name)
+        {
+        }
+
+        void on_tile_end(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y)
+        {
+            for (size_t i = 0, e = frame.aov_images().size(); i < e; ++i)
+            {
+                const string aov_name = frame.aov_images().get_name(i);
+
+                if (aov_name == m_diagnostic_name)
+                {
+                    const Image& image = frame.aov_images().get_image(i);
+                    const Tile& diagnostic_tile = image.tile(tile_x, tile_y);
+                    Tile& aov_tile = m_image.tile(tile_x, tile_y);
+
+                    aov_tile.copy_from(diagnostic_tile);
+
+                    return;
+                }
+            }
+        }
+
+      private:
+        Image&              m_image;
+        const string        m_diagnostic_name;
+    };
+
+    //
+    // Diagnostic AOV.
+    //
+
+    class DiagnosticAOV
+      : public AOV
+    {
+      public:
+        explicit DiagnosticAOV(const char* name, const ParamArray& params)
+          : AOV(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        size_t get_channel_count() const override
+        {
+            return 4;
```

Comment:
Do we need 4 channels?
Some of the previous AOVs use only one (you can save them using 1 channel called "Y") and the pixel variation is RGB (no need for A).

Reply:
Yes I will improve that when implementing per-aov post process